crates/ink/codegen/src/generator/metadata.rs

```diff
@@ -316,6 +358,55 @@ impl Metadata<'_> {
         }
     }
 
+    /// Generates ink! metadata for the given return type of a constructor.
+    /// If constructor result type is not `Result`,
+    /// the metadata will not display any metadata for return type.
+    /// Otherwise, the return type is `Result<(), E>`.
+    fn generate_constructor_return_type(ret_ty: Option<&syn::Type>) -> TokenStream2 {
+        /// Returns `true` if the given type is `Self`.
+        fn type_is_self_val(ty: &syn::Type) -> bool {
+            matches!(ty, syn::Type::Path(syn::TypePath {
+            qself: None,
+            path
+        }) if path.is_ident("Self"))
+        }
+
+        match ret_ty {
+            None => {
+                quote! {
+                    ::ink::metadata::ReturnTypeSpec::new(::core::option::Option::None)
+                }
+            }
+            Some(ty) => {
+                if type_is_self_val(ty) {
+                    return quote! { ::ink::metadata::ReturnTypeSpec::new(::core::option::Option::None)}
+                }
+                let segments = Self::generate_constructor_type_segments(ty);
+                let type_token = Self::replace_self_with_unit(ty);
+                quote! {
+                    if !::ink::is_result_type!(#type_token) {
```

Comment:
If the return type can only be `Self` or `Result<Self, E>` and you are sure that this is already enforced by some earlier compilation stage, isn't this is a redundant check that can be removed?

See: #1473